spyder_kernels/console/kernel.py

```diff
@@ -540,67 +540,52 @@ def set_matplotlib_conf(self, conf):
-            self._set_config_option(
-                'InlineBackend.figure_format',
-                conf[figure_format_n]
+            self._set_inline_config_option(
+                'figure_formats', conf[figure_format_n]
             )
 
+        rc = {}
         if resolution_n in conf:
-            self._set_mpl_inline_rc_config('figure.dpi', conf[resolution_n])
-
+            rc.update({'figure.dpi': conf[resolution_n]})
         if width_n in conf and height_n in conf:
-            self._set_mpl_inline_rc_config(
-                'figure.figsize',
-                (conf[width_n], conf[height_n])
-            )
-
+            rc.update({'figure.figsize': (conf[width_n], conf[height_n])})
         if fontsize_n in conf:
-            self._set_mpl_inline_rc_config('font.size', conf[fontsize_n])
-
+            rc.update({'font.size': conf[fontsize_n]})
         if bottom_n in conf:
-            self._set_mpl_inline_rc_config(
-                'figure.subplot.bottom',
-                conf[bottom_n]
-            )
+            rc.update({'figure.subplot.bottom': conf[bottom_n]})
+        if rc:
+            self._set_inline_config_option('rc', rc)
 
         if bbox_inches_n in conf:
-            self.set_mpl_inline_bbox_inches(conf[bbox_inches_n])
+            bbox_inches = 'tight' if conf[bbox_inches_n] else None
+            self._set_inline_config_option(
+                'print_figure_kwargs', {'bbox_inches': bbox_inches}
+            )
 
-
-    def set_mpl_inline_bbox_inches(self, bbox_inches):
-        """
-        Set inline print figure bbox inches.
-
-        The change is done by updating the 'print_figure_kwargs' config dict.
-        """
-        config = self.config
-        inline_config = (
-            config['InlineBackend'] if 'InlineBackend' in config else {})
-        print_figure_kwargs = (
-            inline_config['print_figure_kwargs']
-            if 'print_figure_kwargs' in inline_config else {})
-        bbox_inches_dict = {
-            'bbox_inches': 'tight' if bbox_inches else None}
-        print_figure_kwargs.update(bbox_inches_dict)
-
-        # This seems to be necessary for newer versions of Traitlets because
-        # print_figure_kwargs doesn't return a dict.
-        if isinstance(print_figure_kwargs, LazyConfigValue):
-            figure_kwargs_dict = print_figure_kwargs.to_dict().get('update')
-            if figure_kwargs_dict:
-                print_figure_kwargs = figure_kwargs_dict
-
-        self._set_config_option(
-            'InlineBackend.print_figure_kwargs', print_figure_kwargs)
+        # To update rcParams in inline mode we can either do so directly or
+        # re-assert inline mode. However, either will prevent restoring
+        # rcParams when toggling to interactive mode. The workaround is to
+        # first toggle to interactive mode, then back to inline. This updates
+        # the rcParams and restores rcParams when switching to interactive.
```

Comment:
Unfortunately, this is a hack that wouldn't allow users to set their preferred interactive backend using the `%matplotlib` magic. And that would break the workflow of several power users (including myself).

The problem is that with your approach we'd be auto-setting the interactive backend for users with the options saved in Spyder. But we shouldn't be doing that because they can do it with `%matplotlib` too (and you know that once an interactive backend is set in a session, it can't be changed).

Is it not possible to change Matplotlib rc parameters after switching to an interactive backend? Or is the issue that Matplotlib only allows to set rc parameters once per session? (A quick googling [seems to suggest](https://matplotlib.org/stable/api/matplotlib_configuration_api.html#matplotlib.rcdefaults) that rc parameters can be changed on the fly).

Reply:
> Unfortunately, this is a hack that wouldn't allow users to set their preferred interactive backend using the `%matplotlib` magic. And that would break the workflow of several power users (including myself).
> 
> The problem is that with your approach we'd be auto-setting the interactive backend for users with the options saved in Spyder. But we shouldn't be doing that because they can do it with `%matplotlib` too (and you know that once an interactive backend is set in a session, it can't be changed).

Aahh, yes, that is a problem...

> Is it not possible to change Matplotlib rc parameters after switching to an interactive backend? Or is the issue that Matplotlib only allows to set rc parameters once per session? (A quick googling [seems to suggest](https://matplotlib.org/stable/api/matplotlib_configuration_api.html#matplotlib.rcdefaults) that rc parameters can be changed on the fly).

It is possible to set rc parameters after switching to an interactive backend and they can be changed on the fly.

The principle issue is two-fold:
* Manually setting `rcParams` while in inline mode forces these parameters to persist after changing to an interactive backend, resulting unexpected values when switching between interactive and inline modes.
* `rcParams` inherit IPython `InlineBackend` settings only upon activating the inline backend, but re-activating the inline backend while in inline mode also forces these parameters to persist, resulting in unexpected values when switching between interactive and inline modes.

This hack only applies to the circumstance where the user changes Spyder's inline preferences while in inline mode. If I remove the hack, changes to Spyder's inline preferences would not be applied to existing kernels until after changing from, and back to, inline mode. This would not affect other expected functionality and would not preset the interactive backend.

The only other alternative is to explicitly set `rcParams` every time the backend changes: `matplotlib.rc_file_defaults()` on changing to interactive backends; apply Spyder's inline preferences on activating inline mode (or when preferences change while in inline mode). This would require storing the preferences in the `kernel` object and obviates the use of `InlineBackend` settings. I thought it was too heavy-handed and inelegant, but it can nevertheless be done.

Reply:
> The only other alternative is to explicitly set rcParams every time the backend changes

Yep, and I think we should do precisely that. Perhaps it's not as elegant as your current approach but it's more explicit and less hacky.

Reply:
Actually, I don't think this will work either. I can only explicitly set rcParams in spyder-kernels, and toggling the backend via the status bar just sends the magic command directly to the console, which would bypass any mechanism in the kernel to set the rcParam. Furthermore, even if I find a way around this, a user can still explicitly use a magic in the console and bypass the mechanism. Therefore, there will always be an inconsistent update to rcParams.

I think the only way to have consistent, repeatable application of the rcParams is for Spyder to not explicitly set them and rely only on the InlineBackend settings mechanism and accept that changing the inline preferences will not have immediate effect if already in inline mode.

Reply:
> Actually, I don't think this will work either. I can only explicitly set rcParams in spyder-kernels, and toggling the backend via the status bar just sends the magic command directly to the console, which would bypass any mechanism in the kernel to set the rcParam.

Yeah, but the nice thing is that we're already overriding the command that's run when calling `%matplotlib`:

https://github.com/spyder-ide/spyder-kernels/blob/94d107e4fecf4a1b910feb41b96e6f228ab3f9c1/spyder_kernels/console/shell.py#L87-L92

So, you should be able to switch rcParams between inline and interactive before actually setting the backend. And since the shell has access to the kernel (through `self.kernel`) you can leave the methods to do that in the kernel so that they work both when receiving new options from Spyder and using the magic.

Reply:
I didn't realize we were overriding `%matplotlib`; I think that will work perfectly for this purpose. I pushed up new changes.